internal/grpcsync/pubsub_test.go

```diff
@@ -0,0 +1,211 @@
+	pubsub := NewPubSub()
+
+	const numPublished = 10
+
+	ts1 := newTestSubscriber(numPublished)
+	pubsub.Subscribe(ts1)
+	wantMsgs1 := []int{}
+
+	var wg sync.WaitGroup
+	wg.Add(2)
+	// Publish ten messages on the pubsub and ensure that they are received in order by the subscriber.
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			pubsub.Publish(i)
+			wantMsgs1 = append(wantMsgs1, i)
+		}
+		wg.Done()
+	}()
+
+	isTimeout := false
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			select {
+			case <-ts1.onMsgCh:
+			case <-time.After(defaultTestTimeout):
+				isTimeout = true
+			}
+		}
+		wg.Done()
+	}()
+
+	wg.Wait()
+	if isTimeout {
+		t.Fatalf("Timeout when expecting the onMessage() callback to be invoked")
+	}
+	if gotMsgs1 := ts1.receivedMsgs(); !cmp.Equal(gotMsgs1, wantMsgs1) {
+		t.Fatalf("Received messages is %v, want %v", gotMsgs1, wantMsgs1)
+	}
+
+	// Register another subscriber and ensure that it receives the last published message.
+	ts2 := newTestSubscriber(numPublished)
+	pubsub.Subscribe(ts2)
+	wantMsgs2 := wantMsgs1[len(wantMsgs1)-1:]
+
+	select {
+	case <-ts2.onMsgCh:
+	case <-time.After(defaultTestShortTimeout):
+		t.Fatalf("Timeout when expecting the onMessage() callback to be invoked")
+	}
+	if gotMsgs2 := ts2.receivedMsgs(); !cmp.Equal(gotMsgs2, wantMsgs2) {
+		t.Fatalf("Received messages is %v, want %v", gotMsgs2, wantMsgs2)
+	}
+
+	wg.Add(3)
+	// Publish ten messages on the pubsub and ensure that they are received in order by the subscribers.
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			pubsub.Publish(i)
+			wantMsgs1 = append(wantMsgs1, i)
+			wantMsgs2 = append(wantMsgs2, i)
+		}
+		wg.Done()
+	}()
+	errCh := make(chan error, 1)
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			select {
+			case <-ts1.onMsgCh:
+			case <-time.After(defaultTestTimeout):
+				errCh <- fmt.Errorf("")
+			}
+		}
+		wg.Done()
+	}()
+	go func() {
+		for i := 0; i < numPublished; i++ {
+			select {
+			case <-ts2.onMsgCh:
+			case <-time.After(defaultTestTimeout):
+				errCh <- fmt.Errorf("")
+			}
+		}
+		wg.Done()
+	}()
+	wg.Wait()
+	select {
+	case <-errCh:
+		t.Fatalf("Timeout when expecting the onMessage() callback to be invoked")
+	default:
+	}
+	if gotMsgs1 := ts1.receivedMsgs(); !cmp.Equal(gotMsgs1, wantMsgs1) {
+		t.Fatalf("Received messages is %v, want %v", gotMsgs1, wantMsgs1)
+	}
+	if gotMsgs2 := ts2.receivedMsgs(); !cmp.Equal(gotMsgs2, wantMsgs2) {
+		t.Fatalf("Received messages is %v, want %v", gotMsgs2, wantMsgs2)
+	}
+
+	pubsub.Stop()
+
+	go func() {
+		pubsub.Publish(99)
+	}()
+	// Ensure that the subscriber callback is not invoked as instantiated
+	// pubsub has already closed.
+	select {
+	case <-ts1.onMsgCh:
+		t.Fatalf("The callback was invoked after pubsub being stopped")
+	case <-ts2.onMsgCh:
+		t.Fatalf("The callback was invoked after pubsub being stopped")
+	case <-time.After(defaultTestShortTimeout):
+	}
+}
+
+func (s) TestPubSub_PublishMsgs_BeforeRegisterSub(t *testing.T) {
```

Comment:
Is this not already tests in the previous test case [here](https://github.com/grpc/grpc-go/pull/6167/files#diff-a18df39905a826f10c28acd16f9160a16c2bafa86dc3a76dd88d25e5ebc70c77R124)? Maybe remove that check from the previous test case if you would like to have separate test cases for both

Reply:
They have different preconditions.
Previous one is tested by calling `pubsub.Subscribe()` at the status that some test subsribers have already added to `pubsub`.
The other one is tested at the status that no subscriber has added to `pubsub`.

Considering above, should we remove either?